geopandas/io/file.py

```diff
@@ -108,11 +109,27 @@ def to_file(df, filename, driver="ESRI Shapefile", schema=None, **kwargs):
         If specified, the schema dictionary is passed to Fiona to
         better control how the file is written. If None, GeoPandas
         will determine the schema based on each column's dtype
+    index : bool, default None
+        If True, write index into one or more columns (for MultiIndex).
+        Default None automatically determines if index is written if it
+        is either named, is a MultiIndex, or has a non-integer data type.
+
+        .. versionadded:: 0.7
+            Previously the index was not written.
 
     The *kwargs* are passed to fiona.open and can be used to write
     to multi-layer data, store data within archives (zip files), etc.
     The path may specify a fiona VSI scheme.
     """
+    if index is None:
+        # Determine if index attribute(s) should be saved to file
+        index = (
+            list(df.index.names) != [None]
+            or not isinstance(df.index, (pd.RangeIndex, pd.Int64Index))
+            or isinstance(df.index, (pd.DatetimeIndex, pd.TimedeltaIndex))
```

Comment:
Hmm, also PeriodIndex is a Int64Index subclass. Given this complication, maybe we can rather check the exact class instead of using `isinstance` (which allows subclasses); eg like `type(df.index) == pd.Int64Index`

Reply:
I'm also starting to wonder if autodetecting on non-integer is going a bit to far, and that users can always use `index=True`. It's not too late to revert to a simpler two condition autodetect: named or multi-index.

Reply:
In principle, I think a check like `list(df.index.names) != [None] or type(df.index) not in (pd.RangeIndex, pd.Int64Index)` is still a relatively simple check.
Also in explanation it can be relatively simple: "write the index except if it is an integer (Range or Int64Index)  index with no name".

@martinfleis thoughts here?

Reply:
Agree with Joris here. If we are doing the automatic detection, it should cover ideally all possibilities, not just a few. The logic should be - is index meaningful or not?; rather than multiindex/named, that is our internal implementation of such a logic.

Reply:
A `pd.Int64Index` could still be meaningful to an end user, so we definitely should document the exact cases we think are "not meaningful." But, I'd bet most non-range integer indices result from query/subset operations, so they're not often intrinsically interesting. A few common number-like identifiers (e.g. FIPS codes in the US Census) are silently truncated on the left when they have leading zeros, so usually are strings. Elsewhere, non-range integers might be meaningful, so clear documentation is useful. 

```
    index : bool, default None
        If True, write index into one or more columns (for MultiIndex).
        Default None writes the index into one or more columns only if 
        the index is named, is a MultiIndex, or has a non-integer data type.
        If False, no index is written. 
```

Reply:
> A pd.Int64Index could still be meaningful to an end user,

Yes, and an automatic detection will never be perfect. But indeed, if we go with this (it might still provide a good default for the majority of use cases), we need to have good docs about this. 

> But, I'd bet most non-range integer indices result from query/subset operations, so they're not often intrinsically interesting.

That was indeed Martin's reasoning for not wanting to write plain Int64Index by default if there is no name.

Thanks for the doc suggestion!

Reply:
OK, let's go with the automatic detection as discussed (and now done in the PR). @mwtoews if you would just like to update the check with my suggestion above?

Reply:
Hopefully this PR isn't too late for the upcoming 0.7 release! I've updated the type detection based on the suggestion by @jorisvandenbossche  and updated the docstrings based on the suggestion by @ljwolf .